File: testing/backend/integration/integration-tests/src/main/java/com/timsmithsoftware/integration_tests/TestBuilder.java
package com.timsmithsoftware.integration_tests;

import java.io.IOException;
import java.net.http.HttpClient;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;

import org.json.JSONObject;
import com.timsmithsoftware.integration_tests.models.ApiConnection;
import com.timsmithsoftware.integration_tests.models.ApiRequest;
import com.timsmithsoftware.integration_tests.models.ApiResponse;
import com.timsmithsoftware.integration_tests.models.DatabaseChange;
import com.timsmithsoftware.integration_tests.models.DatabaseConnection;
import com.timsmithsoftware.integration_tests.models.TestResult;

public class TestBuilder {
	private ApiConnection apiConnection = new ApiConnection();
	private DatabaseConnection databaseConnection = new DatabaseConnection();
	
	private ApiRequest _request;
	private ApiResponse _expectedResponse;
	private ApiResponse _actualResponse;
	private DatabaseChange _databaseChange;
	
	public TestBuilder withRequest(ApiRequest request) {
		this._request = request;
		return this;
	}
	
	public TestBuilder withExpectedResponse(ApiResponse response) {
		this._expectedResponse = response;
		return this;
	}

	public TestBuilder withExpectedDatabaseChange(DatabaseChange change) {
		this._databaseChange = change;
		return this;
	}

	public TestBuilder call() {
		// make API request
		// create a client
		HttpClient client = HttpClient.newHttpClient();

		// use the client to send the request
		HttpResponse response;
		try {
			response = client.send(_request.request, BodyHandlers.ofString());
			// the response:
			System.out.println(response.body());

			// check API response
			_actualResponse = new ApiResponse(response.statusCode(), new JSONObject(response.body().toString()));
			// check DB change has been done correctly
			return this;
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (InterruptedException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		return this;
	}

	public TestResult toResult() {
		if (_expectedResponse != null && _actualResponse != null) {
			return new TestResult(_expectedResponse.equals(_actualResponse));
		}
		// check DBResults
		return new TestResult(false);
	}

	public TestBuilder build() {
		return this;
	}
}

File: testing/backend/integration/integration-tests/src/main/java/com/timsmithsoftware/integration_tests/TestBuilder.kt
package com.timsmithsoftware.integration_tests

import com.timsmithsoftware.integration_tests.models.*

class TestBuilder(private var apiConnection: IApiConnection, var databaseConnection: Connection) {

    private var _request: IApiRequest? = null
    private var _expectedResponse: ApiResponse? = null
    private var _actualResponse: ApiResponse? = null
    private var _databaseChange: ((DatabaseConnection) -> Boolean)? = null
    private var _testDescription: String? = null
    private var _testResult: TestResult? = null

    fun withRequest(request: IApiRequest?): TestBuilder {
        _request = request
        return this
    }

    fun withExpectedResponse(response: ApiResponse?): TestBuilder {
        _expectedResponse = response
        return this
    }

    fun withExpectedDatabaseChange(change: ((DatabaseConnection) -> Boolean)): TestBuilder {
        _databaseChange = change
        return this
    }

    fun withNoExpectedDatabaseChange(): TestBuilder {
        _databaseChange = { _: DatabaseConnection -> true }
        return this
    }

    fun call(): TestBuilder {
            _request?.let {
                _actualResponse = apiConnection.call(it)
            }
            return this
    }

    fun toResult(): TestResult {
        // TODO move some validation to the call() function
        _databaseChange?.let { _databaseChange ->
            _expectedResponse?.let { expected ->
                _actualResponse?.let { actual ->
                    val passes = expected == actual
                    if (passes) {
                        val dbConn = databaseConnection
                        if(dbConn is DatabaseConnection) {
                            val testPasses = _databaseChange.invoke(dbConn)
                            testPasses.let {
                                if (it) {
                                    _testResult = TestResult.TRUE
                                    return _testResult!!
                                } else {
                                    _testResult = TestResult.FALSE("Test failed due to incorrect DB change")
                                    return _testResult!!
                                }
                            }
                        }
                    } else {
                        _testResult = TestResult.FALSE(
                            "Test failed due to incorrect API response\n" +
                                    "Expected response:" +
                                    "$expected" +
                                    "Actual response: " +
                                    "$actual"
                        )
                        return _testResult!!
                    }
                }
                _testResult = TestResult.FALSE("Test failed due to no API response")
                return _testResult!!
            }
            _testResult = TestResult.FALSE("Please add the expected API change before calling the test")
            return _testResult!!
        }
        _testResult = TestResult.FALSE("Please add the expected database change before calling the test")
        return _testResult!!
    }

    fun build(): TestBuilder {
        return this
    }

    fun withDescription(description: String): TestBuilder {
        _testDescription = description
        return this
    }
}